x-pack/plugins/infra/server/lib/alerting/metric_threshold/metric_threshold_executor.test.ts

```diff
@@ -462,25 +463,78 @@ describe('The metric threshold alert type', () => {
+      resultState.push(await execute('test.metric.3', true, resultState.pop()));
+      expectActionToBeNoData(mostRecentAction(instanceID));
+      resultState.push(await execute('test.metric.1', true, resultState.pop()));
       expect(mostRecentAction(instanceID)).toBe(undefined);
-      expect(mostRecentAction(instanceIdA).id).toBe(FIRED_ACTIONS.id);
-      expect(mostRecentAction(instanceIdB).id).toBe(FIRED_ACTIONS.id);
+      expectActionToBeAlert(mostRecentAction(instanceIdA));
+      expectActionToBeAlert(mostRecentAction(instanceIdB));
     });
     test('sends No Data alerts for the previously detected groups when they stop reporting data, but not the * group', async () => {
-      await execute('test.metric.3', resultState.pop());
+      await execute('test.metric.3', true, resultState.pop());
       expect(mostRecentAction(instanceID)).toBe(undefined);
-      expect(mostRecentAction(instanceIdA).id).toBe(FIRED_ACTIONS.id);
-      expect(mostRecentAction(instanceIdB).id).toBe(FIRED_ACTIONS.id);
+      expectActionToBeNoData(mostRecentAction(instanceIdA));
+      expectActionToBeNoData(mostRecentAction(instanceIdB));
+    });
+    test('does not send individual No Data alerts when groups disappear if alertOnGroupDisappear is disabled', async () => {
+      resultState.push(await execute('test.metric.2', false, resultState.pop()));
+      expect(mostRecentAction(instanceID)).toBe(undefined);
+      expectActionToBeAlert(mostRecentAction(instanceIdA));
+      expectActionToBeAlert(mostRecentAction(instanceIdB));
+      expectActionToBeAlert(mostRecentAction(instanceIdC));
+      await execute('test.metric.1', false, resultState.pop());
+      expect(mostRecentAction(instanceID)).toBe(undefined);
+      expectActionToBeAlert(mostRecentAction(instanceIdA));
+      expectActionToBeAlert(mostRecentAction(instanceIdB));
+      expect(mostRecentAction(instanceIdC)).toBe(undefined);
+    });
+
+    describe('if alertOnNoData is disabled but alertOnGroupDisappear is NOT disabled', () => {
+      const executeWeirdNoDataConfig = (metric: string, state?: any) =>
+        executor({
+          ...mockOptions,
+          services,
+          params: {
+            groupBy: 'something',
+            sourceId: 'default',
+            criteria: [
+              {
+                ...baseNonCountCriterion,
+                comparator: Comparator.GT,
+                threshold: [0],
+                metric,
+              },
+            ],
+            alertOnNoData: false,
+            alertOnGroupDisappear: true,
+          },
+          state: state ?? mockOptions.state.wrapped,
+        });
+
+      test('does not send a No Data alert with the * group, but then reports groups when data is available', async () => {
+        resultState.push(await executeWeirdNoDataConfig('test.metric.3'));
+        expect(mostRecentAction(instanceID)).toBe(undefined);
+        resultState.push(await executeWeirdNoDataConfig('test.metric.3', resultState.pop()));
```

Comment:
I don't really understand what's happening with resultState pushing and popping in these alerts?